### PR TITLE
Sequential Quad Tree Creation

### PR DESCRIPTION
**Pull request by ayushnaique**
Code for Quad Trees, we define a grid using two point structs -> TopLeft and BottomRight 

Defined a function to add points to QuadTree
-> Given a list of points, we add them into the grid
-> The base case of adding point to grid is when the actual size of the grid becomes 1 (i.e. 1 x 1)

Defined a function to Search for any given point based on its co-ordinates

Defined a function to check if the given point is within the bounds of the current grid

### CODE REVIEW

File: quadtree.cpp [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    Point(int xVal = 0, int yVal = 0){
+        x = xVal;
+        y = yVal;
+    }
+};
+
+// The Nodes of a QuadTree
+struct Node {
+    Point NodePos;
+    int data;
```
[the remark this answers is not in the document]
Comment (ayushnaique) [reply quadtree.cpp:20]: Yeah can remove that, I thought that each coordinate entry will contain some data, so added that

File: quadtree.cpp [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    Node(Point p, int d){
+        NodePos = p;
+        data    = d;
+    }
+};
+
+// QUADTREE
+class QuadTree {
+    // Top and Bottom point coordinates for B-Box
+    Point TopLeft;
```
[the remark this answers is not in the document]
Comment (ayushnaique) [reply quadtree.cpp:32]: Can change that, no worries

File: quadtree.cpp [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+public:
+    // Inititalize Tree
+    QuadTree(Point TopL, Point BotR){
+        TopLeft     = TopL;
+        BottomRight = BotR;
+
+        n = NULL;
+
+        TopRightTree    = NULL;
```
[the remark this answers is not in the document]
Comment (ayushnaique) [reply quadtree.cpp:52]: I think NULL and nullptr have the same functionality, (NULL is more generic)

File: quadtree.cpp [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+    Point(int xVal = 0, int yVal = 0){
+        x = xVal;
+        y = yVal;
+    }
+};
+
+// The Nodes of a QuadTree
+struct Node {
+    Point NodePos;
```
[the remark this answers is not in the document]
Comment (ayushnaique) [reply quadtree.cpp:19]: The QuadTree is the grid itself, node is the final point that we enter into the QuadTree, will have a coordinate and value associated to it

File: quadtree.cpp [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    return ((P.x >= TopLeft.x && P.x <= BottomRight.x) && (P.y <= TopLeft.y && P.y >= BottomRight.y));
+}
+
+void QuadTree :: insert(Node* node){
+    // If node is empty
+    if(node == NULL)
+        return;
+
+    // Check if inBoundary, else not possible -> Return
+    if(!inBoundary(node->NodePos))
```
[the remark this answers is not in the document]
Comment (ayushnaique) [reply quadtree.cpp:76]: Can do it, i had a simple implementation going on in my head, we can create a parent QuadTree pointer and add it accordingly.

File: quadtree.cpp [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        return;
+
+    // If we have reached the smallest possible grid size
+    if(abs(TopLeft.x - BottomRight.x) <= 1 && abs(TopLeft.y - BottomRight.y) <= 1){
```
[the remark this answers is not in the document]
Comment (ayushnaique) [reply quadtree.cpp:80]: That is the base case, when we reach the smallest grid possible, this is the stop condition for the recursion

Like the grid will contain multiple indices (here the smallest is 1x1 box) -> Could be changed to (0.n1 x 0.n1)  also

**Comment by adityavinodk** [comment]
I went through this briefly and it looks good. Please test it with the `points.txt` generated by `generate_points.py` and print the time.

File: quadtree.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (anuva98) [reply quadtree.cpp:76]: As discussed we can assume solid bounds.

[the remark this answers is not in the document]
Comment (ayushnaique) [reply quadtree.cpp:76]: Yep, exactly what it is doing

**Comment by adityavinodk** [comment]
Is this PR still relevant or can this be closed?

**Comment by ayushnaique** [comment]
can close it

**Pull request closed without merging**